test/unit/context/test-node-values.js

```diff
@@ -0,0 +1,878 @@
+      expect(calcSpy).to.have.callCount(5);
+
+      calcSpy.resetHistory();
+      parent.values.remove(Concat, 'OWNER2');
+      clock.runAll();
+
+      expect(parentStub).to.be.calledTwice.calledWith('AB');
+      expect(sibling1Stub).to.be.calledTwice.calledWith('ABC');
+      expect(sibling2Stub).to.be.calledTwice.calledWith('AB');
+      expect(grandparentStub).to.be.calledOnce;
+      expect(cousin1Stub).to.be.calledOnce;
+      expect(calcSpy).to.have.callCount(3);
+
+      calcSpy.resetHistory();
+      parent.values.remove(Concat, 'OWNER1');
+      clock.runAll();
+
+      expect(parentStub).to.be.calledThrice.calledWith('A');
+      expect(sibling1Stub).to.be.calledThrice.calledWith('AC');
+      expect(sibling2Stub).to.be.calledThrice.calledWith('A');
+      expect(grandparentStub).to.be.calledOnce; // no change.
+      expect(cousin1Stub).to.be.calledOnce; // no change.
+      expect(calcSpy).to.have.callCount(3);
+    });
+
+    it('should recompute the value on parent change', async () => {
+      grandparent.values.set(Concat, 'OWNER1', 'A');
+      parent.values.set(Concat, 'OWNER1', 'B');
+      sibling1.values.set(Concat, 'OWNER1', 'C');
+      clock.runAll();
+
+      expect(grandparentStub).to.be.calledOnce.calledWith('A');
+      expect(parentStub).to.be.calledOnce.calledWith('AB');
+      expect(sibling1Stub).to.be.calledOnce.calledWith('ABC');
+      expect(sibling2Stub).to.be.calledOnce.calledWith('AB');
+      expect(cousin1Stub).to.be.calledOnce.calledWith('A');
+      expect(calcSpy).to.have.callCount(5);
+
+      calcSpy.resetHistory();
+      tree.appendChild(parent.node);
+      rediscover(parent);
+      clock.runAll();
+
+      expect(parentStub).to.be.calledTwice.calledWith('B');
+      expect(sibling1Stub).to.be.calledTwice.calledWith('BC');
+      expect(sibling2Stub).to.be.calledTwice.calledWith('B');
+      expect(grandparentStub).to.be.calledOnce; // no change.
+      expect(cousin1Stub).to.be.calledOnce; // no change.
+      expect(calcSpy).to.have.callCount(3);
+    });
+
+    it('should stop/resume when disconnected/reconnected', async () => {
+      grandparent.values.set(Concat, 'OWNER1', 'A');
+      parent.values.set(Concat, 'OWNER1', 'B');
+      sibling1.values.set(Concat, 'OWNER1', 'C');
+      clock.runAll();
+
+      expect(grandparentStub).to.be.calledOnce.calledWith('A');
+      expect(parentStub).to.be.calledOnce.calledWith('AB');
+      expect(sibling1Stub).to.be.calledOnce.calledWith('ABC');
+      expect(sibling2Stub).to.be.calledOnce.calledWith('AB');
+      expect(cousin1Stub).to.be.calledOnce.calledWith('A');
+      expect(calcSpy).to.have.callCount(5);
+
+      calcSpy.resetHistory();
+      parent.node.remove();
+      await rediscover(parent);
+      clock.runAll();
+      expect(calcSpy).to.have.callCount(0);
+
+      calcSpy.resetHistory();
+      grandparent.values.set(Concat, 'OWNER1', 'Z');
+      parent.values.set(Concat, 'OWNER1', 'Y');
```

Comment:
By "remove parent", do you mean removal of the actual DOM node? Or removal of the `ContextNode` that tracks it and its values? If the latter - I currently do not have any API for removing a `ContextNode` - it can only be removed with the DOM node. The reason for this - I wanted the identity of the `ContextNode.get(domNode)` to never change after it's requested the first time. If the `ContextNode` is no longer needed and no values are set on it, etc - it's still there, it just does not do any work.

Also, notably, I'm not going to export `ContextNode` API directly. Only the downstream APIs, such as:

```
export function setProp(domNode, prop, owner, value);
export function removeProp(...);
export function subscribe(domNode, [deps], handler);
export function unsubscribe(...);
// and so on
```

So, there really won't be a notion of `ContextNode` visible to any external subsystem. Thus nothing really to remove.

Reply:
@jridgewell I found a way to test what you want. I added a couple of tests for reparenting nodes and recomputation that follows. PTAL.